print arrays of Bool as 0/1

stevengj opened the pull request:
As discussed in #30298, this PR changes the printing of `Bool` arrays to output `0` and `1` instead of `true` and `false`.  Rationale:

* The fact that the values are `Bool` is already printed in the `Array` type.
* `0` and `1` is more compact, and is equally correct since `Bool <: Number`.  Correspondingly, it is arguably more readable for a large number of `Bool` values.
* `true` and `false` values are confusing in an algebraic context where users are expecting numeric arrays.  (This will be exacerbated by #30298 if people start using `I(n)` to construct identity matrices.)

Technically, a `Bool` value is printed as `0` or `1` by this PR whenever the container has set `:typeinfo=Bool` in the `IOContext`, indicating that the `Bool` type has already been printed.   So, for example, `true` and `false` are still printed for `Any[true, false]`.

Example:
```jl
###### before this PR ######
julia> Matrix(I, 3,4)
3×4 Array{Bool,2}:
  true  false  false  false
 false   true  false  false
 false  false   true  false

julia> Diagonal(I, 3)
3×3 Diagonal{Bool,Array{Bool,1}}:
  true    ⋅      ⋅  
   ⋅     true    ⋅  
   ⋅      ⋅     true

###### after this PR ######
julia> Matrix(I, 3,4)
3×4 Array{Bool,2}:
 1  0  0  0
 0  1  0  0
 0  0  1  0

julia> Diagonal(I, 3)
3×3 Diagonal{Bool,Array{Bool,1}}:
 1  ⋅  ⋅
 ⋅  1  ⋅
 ⋅  ⋅  1
```

stevengj committed:
print arrays of Bool as 0/1

stevengj commented:
@lstagner, besides being confusing in an algebraic context, a Fortran style is a non-starter because it doesn't parse.  `eval(Meta.parse("Bool[0, 1]"))` works just fine, in contrast.

ararslan commented:
The AppVeyor failure is #30539.

My objection to this is that `Bool`, despite being a `Number`, has its own characteristics, e.g. being the only thing that can be put in a conditional. Printing it as an integer seems to suggest that integers could be used in those contexts, which (thankfully) they cannot. While I agree that getting an array of `true`s and `false`s looks weird for linear algebra, I think this has the potential to be _more_ confusing from a language learning standpoint.

StefanKarpinski commented:
Ahh...
```jl
julia> rand(Bool, 10, 10)
10×10 Array{Bool,2}:
 0  1  0  0  0  1  1  0  1  0
 0  1  0  0  1  0  0  0  1  1
 0  1  0  1  1  0  0  0  1  0
 1  0  1  0  1  1  0  1  0  1
 1  0  1  1  1  1  1  0  1  1
 0  1  1  0  0  1  0  0  1  1
 0  0  0  1  0  1  1  1  1  1
 1  0  0  1  0  1  0  0  0  1
 1  0  0  1  1  1  0  1  0  1
 1  1  1  1  0  0  0  0  1  0
```
That's so much less noisy than before 😁

ararslan commented:
If anything, I think that example further underscores the confusion: You're _asking_ for `Bool` and getting out 1s and 0s, which are not `Bool`s.

StefanKarpinski commented:
Bools are 1s and 0s. They are just 1s and 0s of a particular bit size which, like various other types of integers in Julia, have special input syntax and are printed differently in some context but not others.

ararslan commented:
I mean that they aren't `Bool`s in the sense of https://github.com/JuliaLang/julia/pull/30575#issuecomment-451285269.

StefanKarpinski commented:
But your claim was that Bools aren't 1s and 0s, which is not true. Yes, the bare literals `1` and `0` are not Bools, but they equally aren't Int8s and no one objects to printing `Int8(1)` and `Int8(0)` as "1" and "0". This is precisely the same thing.

ararslan commented:
I misspoke; I've corrected the comment. I mean that literal ones and zeros cannot be used like `Bool`s.

> but they equally aren't Int8s and no one objects to printing Int8(1) and Int8(0) as "1" and "0". This is precisely the same thing.

My point is that it isn't the same thing, because `Bool`s are special in that they're the only things that can be used in conditionals. So printing a `Bool` as 0 or 1 suggests that they're interchangeable in a way that they are not.

stevengj commented:
I doubt that printing a Boolean array is something that usually comes up in the first few minutes of using Julia (except for linear algebra demos where one might print `I(n)` early on!). Scalar boolean printing is not changed here. 

In consequence, I would lean toward optimizing the printing of Bool arrays for slightly more experienced users.

StefanKarpinski commented:
I tend to think that this will likely help people understand how Bools work in Julia by helping them realize that Bools are numbers and that `false == 0` and `true == 1`.

stevengj committed:
performance improvement

Before this PR, with `x = rand(Bool, 10^6)`, `@time repr(x)` took `0.067077 seconds (43 allocations: 8.003 MiB)`.  With the previous version of this PR, I got `0.230414 seconds (2.00 M allocations: 95.556 MiB, 4.60% gc time)` due to the type instability.  After this commit I get `0.047411 seconds (43 allocations: 4.003 MiB)`.

rfourquet commented on test/show.jl:
```diff
@@ -1432,3 +1432,8 @@ replstrcolor(x) = sprint((io, x) -> show(IOContext(io, :limit => true, :color =>
 # issue #30303
 @test repr(Symbol("a\$")) == "Symbol(\"a\\\$\")"
+
+# printing of bools and bool arrays
```
As there is more than one test, maybe use `@testset` ?

stevengj replied:
already done.

rfourquet commented on test/show.jl:
```diff
@@ -1432,3 +1432,8 @@ replstrcolor(x) = sprint((io, x) -> show(IOContext(io, :limit => true, :color =>
+# printing of bools and bool arrays
+@test repr(true) == "true"
+@test repr(Number[true, false]) == "Number[true, false]"
+@test repr([true, false]) == "Bool[1, 0]"
```
Maybe add a test for `BitArray` ?

stevengj replied:
Will do

rfourquet reviewed (approved):
I have wanted this change for a long time, didn't expect it would be popular!

stevengj committed:
testset, bitvector test, test_repr

stevengj committed:
Merge branch 'boolprint' of git://github.com/stevengj/julia into boolprint

pablosanjose commented:
I personally have no problem with this PR, but if the problem is that 0 and 1 are `Int` literals instead of `Bool`s, one could always `show` using ⚬ and |, for example, or some other Unicode characters resembling 1 and 0?

stevengj commented:
@pablosanjose, see my comment above: https://github.com/JuliaLang/julia/pull/30575#issuecomment-451285182 — we want to preserve the property that `repr` parses and evaluates.

stevengj committed:
Merge branch 'master' into boolprint

stevengj committed:
rm whitespace

stevengj committed:
more whitespace

stevengj committed:
doctest updates

stevengj commented:
Whoops, had to update all of the doctests with bool-array output.

timholy commented:
How about
```julia
julia> bitrand(2,3)
2×3 BitArray{2}:
 t  t  t
 f  t  t
where t=true, f=false
```

ararslan commented:
I kind of like that proposal, but the letters `t` and `f` are unfortunately not very visually distinct.

timholy commented:
I supposed one could do 0 and 1 with `where 0=false, 1=true`

StefanKarpinski commented:
I don't really see why the `where 0=false, 1=true` part is needed. It seems quite obvious and clear that `0` means false and `1` means true.

stevengj commented:
@timholy, I will also raise the same objection to your proposal that I did to the ones above:
```jl
julia> repr(bitrand(2,3))
"Bool[0 0 1; 1 0 1]"
```
parses/evals, whereas yours does not.

stevengj commented:
Besides, even non-programmers understand that true and false on a computer are 0 and 1; [it's even part of pop culture](https://www.youtube.com/watch?v=_fGujzulsas).

StefanKarpinski commented:
I’m gonna merge this soon if someone doesn’t beat me to it.

andyferris commented:
For the record, I would slightly prefer we use a dedicated `UInt1` for numeric/algebraic purposes (with this kind of printing, I suppose, rather than `0x0` and `0x1`). Mostly for the reasons @ararslan expouses... a `Bool` has special language powers that (other) `Number`s don't - powers which (possibly counter-intuitively) are lost via `promote` as well naively evaluating the printing in this PR.

I would also like to see us appreciate and promote linear algebra and data tables on a more equal footing in methods in `Base` - because I believe if we want the language to grow and spread we need to be compellingly better than e.g. both numpy *and* pandas. Would it seem natural for `DataFrame`s to pretty-print a boolean column as `0` and `1` for the reason of compactness? What about built-in `AbstractArray`s that can be viewed as tables rather than linear algebra objects, like `Vector{<:NamedTuple}` and `NamedTuple`s of `Vector`s?

I *do* definitely appriecaite the compactness (even in the `true`/`false` flavoured contexts), and it would be very nice to get rid of the `false`/`true` printing in clearly linear-algebraic contexts.

stevengj commented:
@andyferris, printing as 0/1 is only enabled if the context has set `:typeinfo=>Bool`, i.e. in a context where the user has *already* been told to expect boolean values.  In such a context, the meaning of 0 and 1 should be clear.

For `DataFrame` printing, since the `Bool` type is in the column header, it indeed [sets the `:typeinfo` property](https://github.com/JuliaData/DataFrames.jl/blob/eebef64f10f3c29fb763536a066d34c7fad8b34c/src/abstractdataframe/show.jl#L26-L27), and so a boolean column will be printed like this:
```jl
julia> DataFrame(logic = rand(Bool, 10))
10×1 DataFrame
│ Row │ logic │
│     │ Bool  │
├─────┼───────┤
│ 1   │ 1     │
│ 2   │ 1     │
│ 3   │ 1     │
│ 4   │ 0     │
│ 5   │ 0     │
│ 6   │ 0     │
│ 7   │ 0     │
│ 8   │ 0     │
│ 9   │ 1     │
│ 10  │ 1     │
```
What is the problem with this? Since `Bool` is printed at the top of the column, won't the meaning of the `0` and `1` values be clear?

StefanKarpinski commented:
I get that some people are not 100% sold on this, which is why we are merging it early in the 1.2 cycle as an experiment to see how it feels for a few months. If it causes more confusion than it makes things better, it is easy to revert. The symbols `⚬` and `|` are definitely NOT more universally understood as boolean values than `0` and `1`—people talk about binary 1s and 0s all the time. I cannot help feeling that if I'd made this change in 2011, you all would be used to it and it would never occur to you that there's anything objectionable about printing bools as 0s and 1s in contexts where the type is clear.

JeffBezanson commented:
The benefit of this is clear to me as soon as you look at an array with more than about 4 elements. A grid of 1s and 0s is *far* more readable.

It is not that big a deal that different integer types have some minor different properties. They even have different arithmetic, overflowing at different points. That only Bools are allowed in conditions doesn't seem like a reason to print them more verbosely in all contexts. If Bools were so different from integers that `Bool(0)` didn't work, I could understand the objections. But it does work, so this is really just the same as how we use `typeinfo` for all other integer types. :+1:  to merging this.

andyferris commented:
> What is the problem with this?

I'm not at all certain it is a problem, but it seems completely symmetrical to me that a new linear algebra person may be surprised by the current printing (I think this is the original motivation?) and that a new dataframes person may be surprised by the new printing.

I suspect the compactness of the new printing is sufficient to break the symmetry. As you say, it's also possible for `DataFrames` to control printing but not so much for `LinearAlgebra`, which mostly relies on `Base` for this. So, yeah, why not?

o314 commented:
Would be nice if an optional truechar,falsechar pairs could be supplied.
This could permit to render something like this

```julia
julia> rand(Bool, 10, 10)
10×10 Array{Bool,2}:
 -  x  -  -  -  x  x  -  x  -
 -  x  -  -  x  -  -  -  x  x
 -  x  -  x  x  -  -  -  x  -
 x  -  x  -  x  x  -  x  -  x
 x  -  x  x  x  x  x  -  x  x
 -  x  x  -  -  x  -  -  x  x
 -  -  -  x  -  x  x  x  x  x
 x  -  -  x  -  x  -  -  -  x
 x  -  -  x  x  x  -  x  -  x
 x  x  x  x  -  -  -  -  x  -
```
 
Not only useful for playing naval battle game, but to convey easy to read info to business people.
The principle of double entry table, a generalization of contengency table, pivot table.

This behavior may however better belong to the julia `print` family than to the `show` part.

So may be a complementary question is : is there anyway to call a print(Array{Bool}) to get this?

stevengj commented:
@o314, if you want that display for `A = rand(Bool, 10, 10)`, you can do
```jl
julia> [a ? Text('-') : Text('x') for a in A]
10×10 Array{Text{Char},2}:
 -  -  -  -  -  x  x  -  x  x
 -  x  x  -  -  -  -  -  x  x
 x  x  x  x  x  -  -  -  -  -
 -  x  -  -  x  x  -  x  -  x
 -  -  x  -  x  -  -  x  -  x
 x  x  -  x  x  x  x  -  x  -
 x  x  x  x  x  -  -  -  x  x
 -  x  -  x  -  -  x  x  x  x
 -  x  -  -  x  x  x  x  -  -
 -  -  x  -  -  x  -  -  x  -
```

o314 commented:
This could have be done easily with a comprehension of course,
but reinforcing interface is a great purpose too.

Second point Text does not appear clearly at http://docs.julialang.org, only the macro @text_str is stated.
Should we consider it as a stable api in the future?
So with the information accessible there, we could have done

```julia
[a ? text"-" : text"x" for a in A]
A = rand(Bool, 3,3)
```

But that lead to an overwhelming proliferation of api/syntax around the io landscape.
IMHO multiplying syntax shortcut is not a good practice.

As a side effect/near effect, i'm fighting these hours to get a decent table display (think psql or dataframe) without having to load all the dataframe stuff. It will take hours to locate the code - and as much, i guess to cut excessive dependency too.

A question of timeframe i hope

o314 commented:
Thanks for your feedback.

karajan9 commented:
> 
> 
> I get that some people are not 100% sold on this, which is why we are merging it early in the 1.2 cycle as an experiment to see how it feels for a few months. If it causes more confusion than it makes things better, it is easy to revert.

(Without wanting to dive into the pros and cons of this particular proposal...)
I'm wary of this idea in the context of making things more difficult for beginners. The people who would be most confused by the change likely won't be hit by it until 1.2 is out in the world and it is "too late" to change back.
This approach has a similar feel to me to the global-scope-thing where almost all of the feedback came too late because the number of beginners of Julia using nightly builds is quite small.
Of course this change in no way has the same implications nor would it be terribly difficult to change back -- still I think in principle this course of action has a good likelihood to miss its mark.

JeffBezanson commented:
More to the point, I would say printing booleans as 0 and 1 is just not confusing, period.

I'm not sure what alternative you're suggesting --- never make a change that could possibly "confuse" a beginner, because there's no way to get data on it until it's too late?

stevengj commented:
@karajan9, see also [my comment above](https://github.com/JuliaLang/julia/pull/30575#issuecomment-451336166).

StefanKarpinski commented:
We appreciate the feedback. Personally, I find this change a breath of fresh air every time I encounter it. I didn't realize how much those long, verbose `true` and `false` literals in arrays were bothering me. I keep doing `rand(Bool, 10, 10)` just to see the lovely, terse, clear result. As I've stated in this thread, I think that this provides an excellent teaching moment for users who may be momentarily confused: in Julia, `Bool <: Integer` with `true == 1` and `false == 0`. If that were not the case, then I agree this would be a terrible change. However, since that is the case this change is not only much better for display but it also teaches the user something important about how the language works.

StefanKarpinski commented:
Also note that @stevengj, who is very strongly in favor of this change and has more experience teaching students Julia than anyone (with the possible exception of @alanedelman), is the _only_ person who pushed back on the global scope change. Take that for what it's worth, but I would be a lot more worried about this if both @stevengj and @JeffBezanson weren't also fully in favor.

karajan9 commented:
Thanks for the replies! As I said the comment wasn't about this change in particular. Not only because I think I personally will like the change very much but also because I'm not really sure myself it complicates matters that much (and finally, as Jeff hinted to, because more drastic methods like discussing this to death, releasing test versions or something similar are not really warranted here).

I think nobody in this thread suggested they, personally, wouldn't appreciate the change but they were worried about the consequences for beginners of Julia or programming in general. My point was: if _that's_ the concern an approach like "let's just try it and see if someone objects" isn't really practicable and might come across as neglecting of the issue. To avoid a worst case outcome of everybody screaming bloody murder after the fact, statements like
> As I've stated in this thread, I think that this provides an excellent teaching moment for users who may be momentarily confused: in Julia, `Bool <: Integer` with `true == 1` and `false == 0`. [...]

and 
> Also note that @stevengj, who is very strongly in favor of this change and has more experience teaching students Julia than anyone [...]

are much more helpful and satisfactory, I think.

tkf commented:
`rand(Bool, 10, 10)` looks really nice.  However, I find the following a bit confusing:

```julia
julia> [(true, 0), (false, 1), (true, 2), (false, 3)]
4-element Array{Tuple{Bool,Int64},1}:
 (1, 0)
 (0, 1)
 (1, 2)
 (0, 3)
```

Also:

```julia
julia> [true => false]
1-element Array{Pair{Bool,Bool},1}:
 1 => 0

julia> Dict(true => false)  # inconsistent with above?
Dict{Bool,Bool} with 1 entry:
  true => false

julia> show(Dict(true => false))  # needs Dict{Bool,Bool}?
Dict(1=>0)
```

It looks like `display(Dict(...))` does not set `typeinfo` while `show(Dict(...))` does.

StefanKarpinski commented:
Definitely needs to be smoothed out a bit.

diegozea commented:
I've found this because it makes this test fail on nightly: https://github.com/diegozea/MIToS.jl/blob/a8c4075673bf439382c10592211e778f75527b7f/test/Scripts/Template.jl#L47

Maybe it will be better to define a binary type for LinearAlgebra instead of using Bool. This is at least confusing...
